text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,369 @@
+- Feature Name: `nit-lint-level`
+- Start Date: 2024-11-15
+- RFC PR: [rust-lang/rfcs#3730](https://github.com/rust-lang/rfcs/pull/3730)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a new visible lint level below `warn` to allow linters to act like a pair-programmer / code-reviewer where feedback is evaluated on a case-by-case basis.
+- `cargo` does not display these lints by default, requiring an opt-in
+- This RFC assumes LSPs/IDEs will opt-in
+- CIs could opt-in and open issues in the code review to show nits introduced in the current PR but not block merge on them
+
+There is no expectation that crates will be `nit`-free.
+
+*Note:* The name `nit` is being used for illustrative purposes and is assumed to not be the final choice
+
+*Note:* This RFC leaves the determination of what lints will be `nit` by
+default to the respective teams.  Any lints referenced in this document are
+for illustrating the intent of this feature and how teams could reason about
+this new lint level.
+
+# Motivation
+[motivation]: #motivation
+
+By its name and literal usage, the `warn` level is non-blocking.
+However, most projects treat `warn` as a soft-error.
+It doesn't block for local development but CI blocks it from being merged.
+This convention is not new with the Rust community; many C++ projects have take
+this approach before Rust came to be with "warnings clean" being a goal for
+correctness.
```

Comment:
```suggestion
correctness. In addition, many Rust users push for warning-free code based on
bad experiences with past projects that pervasively ignored warnings.
```

Reply:
While its fairly minor change, I'm not sure what this change would buy us.  It starts off talking about the Rust community and then brings up C++ to show that this has a longer history / prior art.

Bringing back up Rust projects doing this seems redundant.

Reply:
doesn't that just give nit fatigue?

Like fundamentally if there's a lot of code you're in charge of, and it's doing something we don't detect, then we detect it, you'll have a lot to fix. and it doesn't matter if we call that new check a nit or a warning. if it suddenly says there's 4000 "whatever" in your project, you'll sigh deeply.

Reply:
The intention of this is that maintainers don't feel the need to know how many "whatever"s they have in a project.  Ideally, both the code author (though LSP) and the code reviewer (through code review integration) see the feedback and decide whether it should be acted on. Or put another way, these are only of interest for new code and can be ignored otherwise.

The one exception is if you are using this to adopt a new lint gradually.  In that case you can either fix everything in which case you only care about the numbers for that specific lint going down or you only worry about new code, assuming the existing code is battle tested enough and through evolution will eventually be rewritten in which case you still don't care about the metric.

Reply:
I understand your intent, but absolutely nothing in this RFC actually does that. Nowhere does this RFC actually explain how "new code" is automatically detected, and nits only run against that. Are you assuming that it only scans for nits in the current git diff, or something like that? Such a system needs to be explained in the RFC if you're hanging the design on it.
* What about people not using git?
* what about people not using any revision control at all? does this never show up in the rust playground since that doesn't use version control?

Reply:
@epage This resolves my concern. I'll make a further comment separately about the wording.

Reply:
@Lokathor moved this to a thread at https://github.com/rust-lang/rfcs/pull/3730/files#r1848991532 as this was for a specific wording change which has been resolved and I don't want to lose track of this thread!